MIL example

myron opened the pull request:
This adds an example of MIL classification from WSI images. 

Let's discuss, which components we can move to MONAI core.
I think we can move MILmodel and ImageExtractPatches transform. 
There are more components we can move too (or just keep them here as part of the tutorial)

Nic-Ma commented on 2d_classification/MIL/mil_classify_monai.py:
```diff
@@ -0,0 +1,944 @@
+            self.avg = self.sum / self.count if self.count > 0 else self.sum
+
+
+def distributed_all_gather(tensor_list, out_numpy=False):
```
Hi @myron ,

For this dist function, MONAI has some similar API:
https://github.com/Project-MONAI/MONAI/blob/dev/monai/utils/dist.py#L42

Thanks.

myron replied:
I looked at that function, it's similar but does a log of unnecessary calculations. it checks for the size of the first dimension of tensors, which does some padding and cropping. I just need a simple gather on tensors,  we don't need that overhead

Nic-Ma replied:
Hi @myron ,

When you `all_gather` PyTorch Tensors in distributed program, the Tensors may have different batch size, in that case, we must pad to make sure the first dim is same, otherwise, NCCL will hang for timeout. That's why PyTorch `DistributedSampler` drops / pads data to make data evenly divisible:
https://github.com/pytorch/pytorch/blob/master/torch/utils/data/distributed.py#L37

I think you can use the MONAI API instead, if your data shape are same, it will not pad / crop at all.

Thanks.

myron replied:
my shape is the same. what is the monai API?

Nic-Ma replied:
Hi @myron ,

If your shape is fixed, the line with skip:
https://github.com/Project-MONAI/MONAI/blob/dev/monai/utils/dist.py#L82
So the MONAI API is similar to your function, right?

Thanks.

myron replied:
may be, that function has overhead, creating additional tensors and synchronizing on them
https://github.com/Project-MONAI/MONAI/blob/4c7a1873869a33fbc643e462d4cb167e8e7785ea/monai/utils/dist.py#L78

I want a simple and fast way, I think I can just keep this funciton in the tutorial

Nic-Ma replied:
OK, could you please help compare the speed of the API and your function? If your function is faster, please feel free to use it in the tutorial.
I suspect maybe they have the same speed.

Thanks.

Nic-Ma commented on 2d_classification/MIL/mil_classify_monai.py:
```diff
@@ -0,0 +1,944 @@
+            # transforms.Lambdad(keys=["label"], func=lambda x: (np.arange(num_classes)<int(x)).astype(np.float32)), #encode label (same but can't pickle for multiprocessing in dataloader)
+            ImageExtractPatchesd(keys=["image"], tile_count=max_tiles, tile_size=tile_size, tile_random_offset=True, tile_all=False, tile_background_val=255),
+            BatchedRandFlip(keys=["image"]),
+            transforms.CastToTypeD(keys=["image"], dtype=np.float32),
```
I submitted a PR for the `dtype` of ScaleIntensityRange: https://github.com/Project-MONAI/MONAI/pull/3235.

Thanks.

myron replied:
Thanks

drbeh reviewed (changes requested):
It's better to move this file to `pathology/multi_instance_learning/` folder. Thanks

myron commented:
I've moved it to pathology/multiple_instance_learning folder and updated the code to import existing components from monai (newly added and fixed components/transforms)

myron commented:
There are some utility functions left here, I intend to keep them in the tutorial without adding to monai core: 

- AverageMeter
- distributed_all_gather
- LabelEncodeIntegerGraded
- SimpleTileFlip

@drbeh sounds good?

drbeh commented:
> There are some utility functions left here, I intend to keep them in the tutorial without adding to monai core:
> 
> * AverageMeter
> * distributed_all_gather
> * LabelEncodeIntegerGraded
> * SimpleTileFlip
> 
> @drbeh sounds good?

Thanks @myron! 
At the end, we have to move all this components to MONA but it is about what the most we can add into MONAI for this release, so let's see what else we can implement till Nov 18. 
`LabelEncodeIntegerGraded` seems to be an important piece, isn't it?

myron commented:
LabelEncodeIntegerGraded - a way to embed class (when a class is a relative tumor grade).  I've not run into it much previously, so it's probably not very common.  We could  consider it adding it

myron commented:
I've updated the tutorial

- using collate function to reassemble data (from list of dicts)
- adjustments for the changes in transforms

drbeh reviewed (changes requested):
@myron could you please remove `datalist_panda_0.json`, upload it [here](https://drive.google.com/drive/u/0/folders/1CAHXDZqiIn5QUfg5A7XsK1BncRu6Ftbh), and add download instruction in the notebook:
```
dataset_url = "https://drive.google.com/uc?id=........."
dataset_path = "datalist_panda_0.json"
gdown.download(dataset_url, dataset_path, quiet=False)
```

Nic-Ma commented on pathology/multiple_instance_learning/README.md:
````diff
@@ -0,0 +1,108 @@
+    --validate
+```
+
+## Questions and bugs
````
I feel this `Questions and bugs` may not be necessary in this README doc? CC @wyli 

Thanks.

myron replied:
I saw it in other tutorials, should I remove it?

Nic-Ma replied:
OK, it's fine to keep it maybe users don't want to check MONAI overall doc clearly.

Thanks.

Nic-Ma commented on pathology/multiple_instance_learning/panda_mil_train_evaluate_pytorch_gpu.py:
```diff
@@ -0,0 +1,546 @@
+        args.world_size = ngpus_per_node * args.world_size
+
+        print("Multigpu", ngpus_per_node, "rescaled lr", args.optim_lr)
+        mp.spawn(main_worker, nprocs=ngpus_per_node, args=(args,))
```
I remember `torch.distributed.launch` can provide better performance for multi-GPU training, should we change it?
CC @vfdev-5 

Thanks.

vfdev-5 replied:
I do not have measurements right now, but I remember checking that some time ago and in recent pytorch versions the difference is almost invisible. I think you can keep it.

Nic-Ma replied:
Hi @vfdev-5 , thanks for your confirmation.

myron replied:
sounds like we can keep it

Nic-Ma commented on pathology/multiple_instance_learning/panda_mil_train_evaluate_pytorch_gpu.py:
```diff
@@ -0,0 +1,546 @@
+    if args.rank == 0:
+        print("Dataset training:", len(dataset_train), "validation:", len(dataset_valid))
+
+    # os.environ["TORCH_HOME"] = "../../torchhome" #TODO, remove later
```
Could you please help remove according to the comment?

Thanks.

myron replied:
ok

Nic-Ma reviewed (commented):
Thanks for the quick update, I feel it overall looks good to me now, just several minor comments inline.
@drbeh , I saw you set `request change`, do you have any more comments?

Thanks in advance.

Nic-Ma commented on pathology/multiple_instance_learning/panda_mil_train_evaluate_pytorch_gpu.py:
```diff
@@ -0,0 +1,546 @@
+import os
```
I feel this file name is not very clear, maybe change it to `train_evaluate.py`?

Thanks.

myron replied:
we have a similar name for another pathology tutorial.  I think it should have at least some distinct info in the name.

drbeh commented on pathology/multiple_instance_learning/panda_mil_train_evaluate_pytorch_gpu.py:
```diff
@@ -0,0 +1,546 @@
+
+    parser = argparse.ArgumentParser(description="Multiple Instance Learning (MIL) example of classification from WSI.")
+    parser.add_argument(
+        "--data_root", default="/PandaChallenge2020/train_images/", help="path to root folder of images"
```
Here and elsewhere, can you change the underscore (`_`) in option names to dash (`-`)? It is easier (and more common way) to pass commandline args with dashes and argparse will automatically convert them to underscore for the variable name.
```suggestion
        "--data-root", default="/PandaChallenge2020/train_images/", help="path to root folder of images"
```

myron replied:
argparse doesn't work well with dashes in names, it will be very hard to use it
https://stackoverflow.com/questions/12834785/having-options-in-argparse-with-a-dash

drbeh replied:
@myron What you are referring to is positional arguments but here you have optional arguments starting with `--`. For `argparse` "Any internal - characters will be converted to _ characters to make sure the string is a valid attribute name."
https://docs.python.org/dev/library/argparse.html#dest

myron replied:
I prefer to keep underscores in the name, it's explicit, and there will be no conversion between the definition and internal use.  I think it's perfectly fine to keep underscores

drbeh replied:
Alright!

drbeh commented on pathology/multiple_instance_learning/panda_mil_train_evaluate_pytorch_gpu.py:
```diff
@@ -0,0 +1,546 @@
+                    logits = calc_head(logits)
+
+                else:
+                    # if number of instances is not big
+                    # we can run inference directly
```
make it one line please.

myron replied:
ok

drbeh commented on pathology/multiple_instance_learning/panda_mil_train_evaluate_pytorch_gpu.py:
```diff
@@ -0,0 +1,546 @@
+        main_worker(0, args)
+
+
+# combine instances from a list of dicts into a single dict, by stacking them along first dim
+# [{'image' : 3xHxW}, {'image' : 3xHxW}, {'image' : 3xHxW}...] - > {'image' : Nx3xHxW}
+# followed by the default collate which will form a batch BxNx3xHxW
```
Please put it as the docstring of the function instead of floating comments.

myron replied:
ok

myron commented on pathology/multiple_instance_learning/panda_mil_train_evaluate_pytorch_gpu.py:
```diff
@@ -0,0 +1,546 @@
+    parser.add_argument(
+        "--data_root", default="/PandaChallenge2020/train_images/", help="path to root folder of images"
+    )
+    # parser.add_argument('--dataset_json', default='./datalist_panda_0.json', help="path to dataset json file")
```
ok

myron commented:
I've made the changes

drbeh commented:
@myron can you address the unresolved comments so that we can merge this example?

drbeh reviewed (commented):
Just waiting for https://github.com/Project-MONAI/tutorials/pull/431#discussion_r753246835 to be addressed.

drbeh commented:
@Nic-Ma, the PR looks good to me. Have your comments been addressed? Do you have any other comment? Thanks

Nic-Ma commented:
Paste the comments from Daguang:
1. Do we need “pip install "git+https://github.com/Project-MONAI/MONAI#egg=monai[tifffile,imagecodecs]"”? This is a MONAI tutorial and people should have MONAI installed already.
2. Can you add some figures for the curves? Training loss and validation curve.

Thanks.

drbeh commented:
@myron 

For the dependencies, you can follow the “Install dependencies" section of other pathology notebook: https://github.com/Project-MONAI/tutorials/blob/master/pathology/tumor_detection/ignite/profiling_camelyon_pipeline.ipynb

First, you can check if MONAI is install or not, and then install it only if the user doesn’t have MONAI installed. Also you can do the same for the individual dependencies, so if the user has MONAI installed but not the dependencies, you can install them only if they are not already installed:
```
!python -c "import monai" || pip install -q git+https://github.com/Project-MONAI/MONAI#egg=MONAI[tifffile,imagecodecs]
!python -c "import tifffile " || pip install -q tifffile
!python -c "import imagecodecs " || pip install -q imagecodecs
```

myron commented:
I've added the curves

myron commented:
> @myron
> 
> For the dependencies, you can follow the “Install dependencies" section of other pathology notebook: https://github.com/Project-MONAI/tutorials/blob/master/pathology/tumor_detection/ignite/profiling_camelyon_pipeline.ipynb
> 
> First, you can check if MONAI is install or not, and then install it only if the user doesn’t have MONAI installed. Also you can do the same for the individual dependencies, so if the user has MONAI installed but not the dependencies, you can install them only if they are not already installed:
> 
> ```
> !python -c "import monai" || pip install -q git+https://github.com/Project-MONAI/MONAI#egg=MONAI[tifffile,imagecodecs]
> !python -c "import tifffile " || pip install -q tifffile
> !python -c "import imagecodecs " || pip install -q imagecodecs
> ```

I mean, I know how to do it. But what do you suggest to put in README for users? do you want to add all these instructions (seems pretty verbose and complicated). I think we can keep it as is. Thanks

drbeh commented on pathology/multiple_instance_learning/README.md:
```diff
@@ -0,0 +1,119 @@
+
+## Dependencies and installation
+
+### MONAI
```
All the headings should be surrounded by blank lines before and after.

myron replied:
ok

drbeh commented on pathology/multiple_instance_learning/README.md:
````diff
@@ -0,0 +1,119 @@
+
+### MONAI
+Please install the required dependencies
+```
````
Please define the language here and add a blank line before.
```suggestion

```bash
```

myron replied:
ok

drbeh commented on pathology/multiple_instance_learning/README.md:
````diff
@@ -0,0 +1,119 @@
+pip install imagecodecs
+```
+
+
````
There is an extra blank line here. Please remove it.

myron replied:
ok

drbeh reviewed (commented):
@myron there are few formatting issues that I mentioned. Apart from that it looks good to me.
@Nic-Ma do you have any other comment?

myron commented:
done

Nic-Ma commented:
Hi @drbeh ,

I see you marked `request change` which blocks merging this PR, do you have any more comments?

Thanks.
